compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -311,6 +310,47 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         interp_ok(new_ptr)
     }
 
+    /// If this grows the allocation, `init_growth` determines
+    /// whether the additional space will be initialized.
+    pub fn reallocate_ptr(
+        &mut self,
+        ptr: Pointer<Option<M::Provenance>>,
+        old_size_and_align: Option<(Size, Align)>,
+        new_size: Size,
+        new_align: Align,
+        kind: MemoryKind<M::MemoryKind>,
+        init_growth: AllocInit,
+    ) -> InterpResult<'tcx, Pointer<M::Provenance>> {
+        self.reallocate_ptr_inner(
+            ptr,
+            old_size_and_align,
+            new_size,
+            new_align,
+            kind,
+            kind,
+            init_growth,
+        )
+    }
+
+    pub fn leak_const_heap_ptr(
+        &mut self,
+        ptr: Pointer<Option<M::Provenance>>,
+        size: Size,
+        align: Align,
+        kind: MemoryKind<M::MemoryKind>,
+        new_kind: MemoryKind<M::MemoryKind>,
+    ) -> InterpResult<'tcx, Pointer<M::Provenance>> {
+        self.reallocate_ptr_inner(
```

Comment:
Why does this reallocate? I don't think that is the right semantics. It will be very hard to use this to intern a graph of allocations since you need to mutate the all pointers into the allocation. And it will be impossible to do this in a cyclic graph.

Reply:
> It will be very hard to use this to intern a graph of allocations since you need to mutate the all pointers into the allocation.

True. I originally went with a "just change the original memory kind" approach but got really scared by the following lines:

https://github.com/rust-lang/rust/blob/688ea65df6a47866d0f72a00f1e18b47a7edf83b/compiler/rustc_const_eval/src/interpret/memory.rs#L1609-L1614

Reply:
Hm, I don't see how that comment would conflict with "change the memory kind"?

Reply:
I made it so that it now makes the allocation global (`tcx` managed).